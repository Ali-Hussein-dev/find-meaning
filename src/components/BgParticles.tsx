import ParticleField from 'react-particles-webgl';
const config = {
  showCube: false,
  dimension: '2D',
  velocity: 1.2,
  boundaryType: 'passthru',
  antialias: false,
  direction: {
    xMin: -1,
    xMax: 1,
    yMin: -1,
    yMax: 1,
    zMin: -1,
    zMax: 1,
  },
  lines: {
    colorMode: 'rainbow',
    color: '#351CCB',
    transparency: 0.9,
    limitConnections: true,
    maxConnections: 20,
    minDistance: 110,
    visible: true,
  },
  particles: {
    colorMode: 'rainbow',
    color: '#3FB568',
    transparency: 0.9,
    shape: 'circle',
    boundingBox: 'canvas',
    count: 60,
    minSize: 20,
    maxSize: 50,
    visible: true,
  },
  cameraControls: {
    enabled: false,
    enableDamping: true,
    dampingFactor: 0.2,
    enableZoom: false,
    autoRotate: false,
    autoRotateSpeed: 0.3,
    resetCameraFlag: true,
  },
};
//=======================
export const BgParticles: React.FC = () => {
  // hooks
  //--------------------------------------
  // functions
  //--------------------------------------
  return <ParticleField config={config} />;
};
