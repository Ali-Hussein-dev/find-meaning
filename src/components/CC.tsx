import * as React from 'react';

interface Div extends React.ComponentPropsWithoutRef<'div'> {
  isTrue: any;
  fallback?: React.ReactElement;
}
//======================= CC stands for conditional component
export const CC: React.FC<Div> = ({
  isTrue,
  children,
  fallback = null,
  ...props
}) => {
  return isTrue ? <div {...props}>{children} </div> : fallback;
};
