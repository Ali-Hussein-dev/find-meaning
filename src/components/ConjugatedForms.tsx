import { Tooltip } from '@chakra-ui/react';
import * as React from 'react';
//=======================
type formLabels =
  | 'Infinitive'
  | 'Simple past'
  | 'Past participle'
  | 'Continous';
export type ConjugatedFormsT = [formLabels, string][];

//=======================
export const Pipe: React.FC = () => (
  <span className="text-blueGray-300">|</span>
);

export const ConjugatedForms: React.FC<{
  forms: ConjugatedFormsT;
}> = ({ forms }) => {
  //======================================return
  return (
    <>
      {forms.map((a, i) => (
        <React.Fragment key={i}>
          <Tooltip hasArrow label={[a[0]]} aria-label="a tooltip">
            {a[1]}
          </Tooltip>
          {forms.length - 1 > i ? <Pipe /> : null}
        </React.Fragment>
      ))}
    </>
  );
};
