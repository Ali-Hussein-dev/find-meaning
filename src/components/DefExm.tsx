//=======================
export const DefExm: React.FC<{
  def: string;
  exm: string | string[];
  context?: string[];
}> = ({ def, exm, context }) => {
  return (
    <div className="w-11/12 mb-3 text-blueGray-600">
      <p>
        <span className=" text-lightBlue-500">{context?.join()} </span>
        {def}
      </p>
      <>
        {typeof exm === 'string' ? (
          <p className="ml-1 italic font-light text-trueGray-600">{exm}</p>
        ) : Array.isArray(exm) ? (
          exm?.map((str, i) => (
            <p className="ml-1 italic font-light text-trueGray-600" key={i}>
              {str}
            </p>
          ))
        ) : null}
      </>
    </div>
  );
};
