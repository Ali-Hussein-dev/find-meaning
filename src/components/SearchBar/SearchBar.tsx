import * as React from 'react';
import {
  useSearch,
  Input,
  Delete,
  Suggestions,
  Submit,
  FocusBtn,
} from '@/components/index';
import { isMobile } from 'react-device-detect';

//--------------------------------------
export const SearchCtx = React.createContext(undefined);
//--------------------------------------
export const SearchBar: React.FC = () => {
  const searchbarStore = useSearch();
  const {
    handleSubmit,
    combobox,
    inputRef,
    shouldInputFocus,
    setShouldInputFocus,
  } = searchbarStore;
  const { getComboboxProps } = combobox;
  switch (isMobile) {
    case true:
      return (
        <SearchCtx.Provider value={{ ...searchbarStore }}>
          <div className="flex justify-center w-full bg-blueGray-200">
            <form
              onSubmit={handleSubmit}
              className="flex items-center justify-between w-full border-b border-lightBlue-400 focus-within:border-lightBlue-400 focus-within:border-b-2 bg-blueGray-800"
              {...getComboboxProps({}, { suppressRefError: true })}
            >
              <Input ref={inputRef} />
              <Delete />
              <Submit />
            </form>
            <Suggestions />
          </div>
          <FocusBtn
            shouldInputFocus={shouldInputFocus}
            setShouldInputFocus={setShouldInputFocus}
          />
        </SearchCtx.Provider>
      );
      break;
    default:
      return (
        <SearchCtx.Provider value={{ ...searchbarStore }}>
          <div className="relative flex justify-center w-full ">
            <form
              onSubmit={handleSubmit}
              className="flex items-center justify-between w-full border-b border-lightBlue-400 focus-within:border-lightBlue-400 focus-within:border-b-2 bg-blueGray-800"
              {...getComboboxProps({}, { suppressRefError: true })}
            >
              <Input ref={inputRef} />
              <Delete />
              <Submit />
            </form>
            <Suggestions />
          </div>
        </SearchCtx.Provider>
      );
      break;
  }
};
