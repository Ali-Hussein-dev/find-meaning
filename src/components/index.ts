export * from './SearchBar';
export * from './Pronunciation';
export * from './WikContainer';
export * from './WikiLexeme';
export * from './Gifs';
export * from './CC';
export * from './Input';
export * from './DefExm';
export * from './UrbanContainer';
export * from './BgParticles';
export * from './leaf-components/AsyncStatus';
export * from './leaf-components/HighlightSubstring';
export * from './leaf-components/FeedbackSubmitted';
export * from './FeedbackDrawer';
export * from './Footer';
export * from './leaf-components/CustomSkeleton';
