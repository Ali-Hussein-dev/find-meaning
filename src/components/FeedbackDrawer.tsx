import * as React from 'react';
import { nanoid } from 'nanoid';
import store from 'store2';
import { isMobile } from 'react-device-detect';
import { fetcherPost } from 'src/utils';
import { useQuery } from 'react-query';
import { CondComp, FeedbackSubmitted } from '@/components/index';
import { MdClear, MdSend } from 'react-icons/md';
import {
  Input,
  Textarea,
  Stack,
  useDisclosure,
  Spinner,
  Drawer,
  DrawerBody,
  DrawerFooter,
  DrawerHeader,
  DrawerOverlay,
  DrawerContent,
} from '@chakra-ui/react';

//--------------------------------------setIdentiefer
(function () {
  if (!store.get('identiefer')) {
    store.set('identiefer', { _id: nanoid(), feedbackCount: 0 });
  }
  return;
})();
//--------------------------------------updateIdentieferProperty
const updateIdentieferProperty = (
  identieferObj: {
    _id: string;
    feedbackCount: number;
  },
  keyValue,
) => {
  store.set('identiefer', { ...identieferObj, ...keyValue });
};
const feedbackCount = store.get('identiefer').feedbackCount;
//=======================FeedbackSubmitted

//=======================FeedbackDrawer
export const FeedbackDrawer: React.FC = () => {
  //--------------------------------------hooks
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [nameValue, setNameValue] = React.useState('');
  const [emailValue, setEmailValue] = React.useState('');
  const [textareaValue, setTextareaValue] = React.useState('');
  const [enabledFetching, setEnabledFetching] = React.useState(false);
  const { status } = useQuery(
    'feedback',
    () =>
      fetcherPost('api/feedback', {
        _id: store.get('identiefer')._id,
        feedback: {
          userName: nameValue,
          email: emailValue,
          feedback: textareaValue,
          createdAt: new Date(),
          stars: undefined,
        },
      }),
    {
      enabled: enabledFetching,
      staleTime: Infinity,
      onSuccess: () => {
        updateIdentieferProperty(store.get('identiefer'), {
          feedbackCount: store.get('identiefer').feedbackCount + 1,
        });
      },
    },
  );
  //--------------------------------------functions
  const handleSubmit = async (e) => {
    e.preventDefault();
    setEnabledFetching(true);
  };
  //======================================return
  return (
    feedbackCount < 3 && (
      <div className="w-full mb-3 text-center">
        <Drawer
          isOpen={isOpen}
          placement={isMobile ? 'bottom' : 'right'}
          onClose={onClose}
        >
          <DrawerOverlay>
            <DrawerContent data-testid="feedback-container" bg={''}>
              <DrawerHeader className="bg-lightBlue-800">
                <div className="">
                  <h2 className="text-base text-lightBlue-300">Feedback</h2>
                  <h2 className="text-xl font-normal text-blueGray-50">
                    Your feedback is valuable to us
                  </h2>
                </div>
              </DrawerHeader>
              <DrawerBody className="bg-blueGray-50 text-blueGray-700">
                <form onSubmit={handleSubmit}>
                  <CondComp
                    baseCond={status !== 'success'}
                    fallback={
                      <FeedbackSubmitted msg={'Thank you for your feedback!'} />
                    }
                    className="py-3 "
                  >
                    <Stack spacing={4}>
                      <label htmlFor="name">
                        <Input
                          isRequired
                          id="name"
                          variant="flushed"
                          value={nameValue}
                          onChange={(e) => setNameValue(e.target.value)}
                          type="text"
                          placeholder="name*"
                        />
                      </label>
                      <label htmlFor="email">
                        <Input
                          id="email"
                          isRequired
                          variant="flushed"
                          type="email"
                          placeholder="email*"
                          value={emailValue}
                          onChange={(e) => setEmailValue(e.target.value)}
                        />
                      </label>
                      <label htmlFor="feedback-text">
                        <Textarea
                          id="feedback-text"
                          variant="flushed"
                          isRequired
                          value={textareaValue}
                          onChange={(e) => setTextareaValue(e.target.value)}
                          placeholder="e.g. ask for a nice feature*"
                        />
                      </label>
                    </Stack>
                  </CondComp>
                </form>
              </DrawerBody>
              <DrawerFooter className="flex justify-between w-full gap-x-1 bg-blueGray-50">
                <button
                  type="button"
                  onClick={onClose}
                  className="btn flex-grow h-8 rounded-lg press-effect focus:ring-2 center text-blueGray-600 ring-yellow-600"
                >
                  <span className="center gap-x-2">
                    <MdClear size="20" />
                    Cancel
                  </span>
                </button>
                {status !== 'success' && (
                  <button
                    onClick={handleSubmit}
                    type="submit"
                    name="submit-feedback"
                    data-testid="submit-feedback"
                    className="btn flex-grow h-8 text-white rounded-lg bg-lightBlue-600 disabled:bg-blueGray-200 disabled:text-blueGray-400 disabled:cursor-not-allowed focus:ring-2 center gap-x-1 ring-yellow-600"
                    disabled={
                      !(!!emailValue && !!textareaValue && !!nameValue) ||
                      status === 'loading'
                    }
                  >
                    {status === 'loading' ? (
                      <Spinner size="sm" />
                    ) : (
                      <span className="center gap-x-2">
                        <MdSend size="15" />
                        Send
                      </span>
                    )}
                  </button>
                )}
              </DrawerFooter>
            </DrawerContent>
          </DrawerOverlay>
        </Drawer>
        <button
          type="button"
          onClick={onOpen}
          className="p-1 font-semibold underline rounded focus:outline-none text-blueGray-300 focus:ring-2 ring-lightBlue-500 hover:text-lightBlue-500"
        >
          Was this helpful?
        </button>
      </div>
    )
  );
};
