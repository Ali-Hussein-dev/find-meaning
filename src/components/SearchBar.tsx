import * as React from 'react';
import { BsSearch } from 'react-icons/bs';
import { MdClear } from 'react-icons/md';
import { useRouter } from 'next/router';
import { Spinner } from '@chakra-ui/react';
import { useFetch } from 'src/utils';
import { CondComp, Input } from '@/components/index';

//=======================
export const SearchBar: React.FC = () => {
  // hooks
  const router = useRouter();
  const [inputValue, setInputValue] = React.useState('');
  const [enabledFetching, setEnabledFetching] = React.useState(false);
  // eslint-disable-next-line prefer-const
  let query = inputValue;
  if (typeof router.query.q !== 'string') {
    router.query.q = '';
  }

  const linguaResponse = useFetch(['lingua', query], 'lingua', enabledFetching);
  const { status } = linguaResponse;

  React.useEffect(() => {
    if (typeof router.query?.q === 'string') {
      setInputValue(router.query.q);
    }
    if (enabledFetching) {
      setEnabledFetching(false);
    }
  }, [router.query.q]);
  //--------------------------------------
  // functions
  const handleSubmit = (e) => {
    e.preventDefault();
    if (query.length === 0) {
      console.error('query must be a string and not empty');
    } else {
      setEnabledFetching(true);
    }
    router.push({ query: { q: inputValue.trim().toLowerCase() } });
  };
  //--------------------------------------
  return (
    <div className="w-full pt-5 mb-3 text-trueGray-600">
      <form
        className="flex items-center overflow-hidden border-b rounded-t-lg focus-within:border-lightBlue-400 focus-within:bg-blueGray-700 focus-within:opacity-90 border-blueGray-400 focus-within:border-b-2"
        onSubmit={handleSubmit}
        role="form"
      >
        <Input
          value={inputValue}
          onChange={(e) => setInputValue(e.target.value)}
          disabled={status === 'loading'}
        />
        <div className="flex items-center text-trueGray-500 bg-blueGray-800">
          <CondComp baseCond={inputValue.length > 0}>
            <button
              onClick={() => setInputValue('')}
              disabled={status === 'loading'}
              aria-label="delete-button"
              type="button"
              className="w-12 h-12 rounded-none btn focus:bg-blueGray-500 text-blueGray-200"
            >
              <MdClear size="20" className="mx-auto" />
            </button>
          </CondComp>
          <div className="w-1 h-8 border-r" />
          <span className="grid w-12 h-12 place-items-center">
            <CondComp
              baseCond={status !== 'loading'}
              fallback={<Spinner size="sm" className="text-lightBlue-300" />}
            >
              <button
                disabled={inputValue.length === 0 || status === 'loading'}
                role="button"
                type="submit"
                aria-label="submit"
                className="w-12 h-12 icon-btn focus:bg-blueGray-500 text-blueGray-200"
              >
                <BsSearch className="mx-auto" />
              </button>
            </CondComp>
          </span>
        </div>
      </form>
    </div>
  );
};
