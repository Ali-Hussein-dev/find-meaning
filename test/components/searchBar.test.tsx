// import { mocked } from '@/test/__mocks__';
import { SearchBar } from '@/components/SearchBar';
import { render, screen } from '@/test/testUtils';
import userEvent from '@testing-library/user-event';

// TODO compolete implementation
// beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));

// afterAll(() => server.close());
// afterEach(() => server.resetHandlers());

describe('Search Bar', () => {
  it('should input be disabled WHEN submit event is fired off', () => {
    render(<SearchBar />);
    const submitBtn = screen.getByRole('button', {
      name: /submit/i,
    });
    const inputLabel = screen.getByLabelText(/search-input/i);
    userEvent.type(inputLabel, 'value');
    userEvent.click(submitBtn);
    expect(inputLabel).toBeDisabled();
  });
  //--------------------------------------
  it('should submit btn be disabled WHEN input is still empty', () => {
    render(<SearchBar />);
    const submitBtn = screen.getByRole('button', {
      name: /submit/i,
    });
    const inputLabel = screen.getByLabelText(/search-input/i);
    expect(submitBtn).toBeDisabled();
    userEvent.type(inputLabel, 'v');
    expect(submitBtn).not.toBeDisabled();
  });
  //--------------------------------------
  it('Should del btn displayed WHEN start typing', () => {
    render(<SearchBar />);
    const inputLabel = screen.getByLabelText(/search-input/i);
    let delBtn = screen.queryByRole('button', {
      name: /delete-button/i,
    });
    expect(delBtn).toBe(null);
    userEvent.type(inputLabel, 'd');
    delBtn = screen.getByRole('button', {
      name: /delete-button/i,
    });
    expect(delBtn).toBeInTheDocument();
    expect(delBtn).toBeVisible();
  });
  //--------------------------------------
  it('Should match snapshots ', () => {
    const { container } = render(<SearchBar />);
    expect(container).toMatchSnapshot();
    const inputLabel = screen.getByLabelText(/search-input/i);
    userEvent.type(inputLabel, 'device');
    expect(container).toMatchSnapshot();
  });
  //--------------------------------------
});
