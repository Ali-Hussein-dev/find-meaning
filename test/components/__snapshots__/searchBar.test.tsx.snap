// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar Should match snapshots  1`] = `
<div>
  <div
    class="w-full pt-5 mb-3 text-trueGray-600"
  >
    <form
      class="flex items-center overflow-hidden border-b rounded-t-lg focus-within:border-lightBlue-400 focus-within:bg-blueGray-700 focus-within:opacity-90 border-blueGray-400 focus-within:border-b-2"
      role="form"
    >
      <label
        for="search-input"
      />
      <input
        aria-label="search-input"
        class="w-full h-12 py-1 pl-3 text-lg disabled:text-trueGray-400 disabled:cursor-not-allowed focus:outline-none bg-blueGray-800 text-blueGray-200 focus:opacity-90 focus:bg-blueGray-700"
        name="search-input"
        placeholder="start searching..."
        type="text"
        value=""
      />
      <div
        class="flex items-center text-trueGray-500 bg-blueGray-800"
      >
        <div
          class="w-1 h-8 border-r"
        />
        <span
          class="grid w-12 h-12 place-items-center"
        >
          <div
            data-testid="cc-container"
          >
            <button
              aria-label="submit"
              class="w-12 h-12 icon-btn focus:bg-blueGray-500 text-blueGray-200"
              disabled=""
              role="button"
              type="submit"
            >
              <svg
                class="mx-auto"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10.442 10.442a1 1 0 011.415 0l3.85 3.85a1 1 0 01-1.414 1.415l-3.85-3.85a1 1 0 010-1.415z"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M6.5 12a5.5 5.5 0 100-11 5.5 5.5 0 000 11zM13 6.5a6.5 6.5 0 11-13 0 6.5 6.5 0 0113 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
             
          </div>
        </span>
      </div>
    </form>
  </div>
</div>
`;

exports[`Search Bar Should match snapshots  2`] = `
<div>
  <div
    class="w-full pt-5 mb-3 text-trueGray-600"
  >
    <form
      class="flex items-center overflow-hidden border-b rounded-t-lg focus-within:border-lightBlue-400 focus-within:bg-blueGray-700 focus-within:opacity-90 border-blueGray-400 focus-within:border-b-2"
      role="form"
    >
      <label
        for="search-input"
      />
      <input
        aria-label="search-input"
        class="w-full h-12 py-1 pl-3 text-lg disabled:text-trueGray-400 disabled:cursor-not-allowed focus:outline-none bg-blueGray-800 text-blueGray-200 focus:opacity-90 focus:bg-blueGray-700"
        name="search-input"
        placeholder="start searching..."
        type="text"
        value="device"
      />
      <div
        class="flex items-center text-trueGray-500 bg-blueGray-800"
      >
        <div
          data-testid="cc-container"
        >
          <button
            aria-label="delete-button"
            class="w-12 h-12 rounded-none btn focus:bg-blueGray-500 text-blueGray-200"
            type="button"
          >
            <svg
              class="mx-auto"
              fill="currentColor"
              height="20"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
           
        </div>
        <div
          class="w-1 h-8 border-r"
        />
        <span
          class="grid w-12 h-12 place-items-center"
        >
          <div
            data-testid="cc-container"
          >
            <button
              aria-label="submit"
              class="w-12 h-12 icon-btn focus:bg-blueGray-500 text-blueGray-200"
              role="button"
              type="submit"
            >
              <svg
                class="mx-auto"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10.442 10.442a1 1 0 011.415 0l3.85 3.85a1 1 0 01-1.414 1.415l-3.85-3.85a1 1 0 010-1.415z"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M6.5 12a5.5 5.5 0 100-11 5.5 5.5 0 000 11zM13 6.5a6.5 6.5 0 11-13 0 6.5 6.5 0 0113 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
             
          </div>
        </span>
      </div>
    </form>
  </div>
</div>
`;
